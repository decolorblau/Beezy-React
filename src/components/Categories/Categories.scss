.formats {
  width: 92vw;
  display: flex;
  overflow-x: scroll;
  max-width: 1536px;
  margin: 0 25px 5px;
}

.format {
  min-width: 120px;
  margin: 2px 2px 6px;
  padding: 3px;
  background-color: #2d2d2d;
  width: 100%;
  height: 35px;
  border-radius: 18px;
  border: none;
  font: 13px "Roboto", sans-serif;
  font-weight: 500;
  letter-spacing: 2px;
  color: rgba(220, 220, 220, 0.9);
}

.formats::-webkit-scrollbar {
  -webkit-appearance: none;
}

.formats::-webkit-scrollbar:vertical {
  width: 7px;
}

.formats::-webkit-scrollbar-button:increment,
.formats::-webkit-scrollbar-button {
  display: none;
}

.formats::-webkit-scrollbar:horizontal {
  height: 3px;
}

.formats::-webkit-scrollbar-thumb {
  background-color: rgba(220, 220, 220, 0.9);
  border-radius: 20px;
}

.formats::-webkit-scrollbar-track {
  border-radius: 10px;
}

@media (min-width: 768px) {
  .formats {
    margin-bottom: 10px;
    margin: 0 25px 10px;
  }

  .format {
    min-width: 160px;
    padding: 8px;
    width: 100%;
    height: 56px;
    border-radius: 28px;
    font: 16px "Roboto", sans-serif;
    font-weight: 600;
    letter-spacing: 3px;
  }
}
