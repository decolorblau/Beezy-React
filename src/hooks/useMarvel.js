import axios from "axios";
import { useCallback, useContext } from "react";
import MarvelContext from "../store/contexts/MarvelContext";

const useMarvel = () => {
  const { offset, setComics, orderBy, limit, setTotal, setComicData } =
    useContext(MarvelContext);

  const urlApi = "http://gateway.marvel.com/v1/public/";

  const getComics = useCallback(async () => {
    try {
      const { data } = await axios.get(`${urlApi}comics`, {
        params: {
          orderBy: orderBy,
          limit: limit,
          offset: offset,
          ts: process.env.REACT_APP_MARVEL_TS,
          apikey: process.env.REACT_APP_MARVEL_APIKEY,
          hash: process.env.REACT_APP_MARVEL_HASH,
        },
      });
      const { data: comics } = data;
      setComics(comics.results);
      setTotal(comics.total);
    } catch (error) {
      return error;
    }
  }, [orderBy, limit, offset, setComics, setTotal]);

  const getComicsByName = useCallback(
    async (search) => {
      try {
        const { data } = await axios.get(`${urlApi}comics`, {
          params: {
            orderBy: orderBy,
            limit: limit,
            offset: offset,
            ts: process.env.REACT_APP_MARVEL_TS,
            apikey: process.env.REACT_APP_MARVEL_APIKEY,
            hash: process.env.REACT_APP_MARVEL_HASH,
            titleStartsWith: search,
          },
        });
        const { data: comics } = data;
        setComics(comics.results);
        setTotal(comics.total);
      } catch (error) {
        return error;
      }
    },
    [limit, offset, orderBy, setComics, setTotal]
  );
  const getComicById = useCallback(
    async (id) => {
      try {
        const { data } = await axios.get(`${urlApi}comics/${id}`, {
          params: {
            ts: process.env.REACT_APP_MARVEL_TS,
            apikey: process.env.REACT_APP_MARVEL_APIKEY,
            hash: process.env.REACT_APP_MARVEL_HASH,
          },
        });
        const { data: comic } = data;
        setComicData(comic.results);
        console.log(...comic.results);
      } catch (error) {
        return error;
      }
    },
    [setComicData]
  );

  return {
    getComics,
    getComicsByName,
    getComicById,
  };
};

export default useMarvel;
