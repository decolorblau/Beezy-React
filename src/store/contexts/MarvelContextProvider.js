import MarvelContext from "./MarvelContext";

const MarvelContextProvider = ({ children }) => {
  return <MarvelContext.Provider>{children}</MarvelContext.Provider>;
};

export default MarvelContextProvider;
