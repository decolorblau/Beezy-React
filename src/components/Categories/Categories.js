import "./Categories.scss";
import useMarvel from "../../hooks/useMarvel";
import PropTypes from "prop-types";
import { useContext } from "react";
import MarvelContext from "../../store/contexts/MarvelContext";

const Categories = ({ formats }) => {
  const {
    getComicsByFormat,
    getComics,
    getComicsByFormatAndByName,
    getComicsByName,
  } = useMarvel();
  const { setTitle, setOffset, searchByName, lastSearch } =
    useContext(MarvelContext);

  const filterByGenre = (format) => {
    setOffset(0);
    if (!searchByName) {
      if (format === "all comics") {
        getComics();
      } else {
        getComicsByFormat(format);
      }
    } else {
      if (format === "all comics") {
        getComicsByName(lastSearch);
      } else {
        console.log(lastSearch);
        getComicsByFormatAndByName(lastSearch, format);
      }
    }
    setTitle(format);
  };

  return (
    <div className="formats">
      {formats.map((format) => (
        <button
          className="format"
          key={format.id}
          onClick={() => filterByGenre(format.format)}
        >
          {format.format.toUpperCase()}
        </button>
      ))}
    </div>
  );
};

Categories.protoTypes = {
  format: PropTypes.array.isRequired,
};

export default Categories;
