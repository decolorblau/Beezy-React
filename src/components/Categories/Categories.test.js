import { fireEvent, render, screen } from "@testing-library/react";
import Categories from "./Categories";
import { BrowserRouter as Router } from "react-router-dom";
import MarvelContextProvider from "../../store/contexts/MarvelContextProvider";
import MarvelContext from "../../store/contexts/MarvelContext";
import ReactTestRenderer from "react-test-renderer";
import { renderHook, act } from "@testing-library/react-hooks";
import useMarvel from "../../hooks/useMarvel";

/* const mockGetComics = jest.fn();

jest.mock("../../hooks/useMarvel.js"); */

describe("Given the component Categories", () => {
  describe("When it receives an array of categories", () => {
    test("Then it should return as many buttons as the array length", () => {
      const formatList = [
        { id: 1, format: "comic" },
        { id: 2, format: "magazine" },
        { id: 3, format: "trade paperback" },
      ];

      render(
        <MarvelContextProvider>
          <Router>
            <Categories formats={formatList} />
          </Router>
        </MarvelContextProvider>
      );

      const formatsRendered = screen.getAllByRole("button");

      expect(formatsRendered).toHaveLength(formatList.length);
    });
    test("Then it should be a button with comic text, another with magazine text an another with trade paperback", () => {
      const formatList = [
        { id: 1, format: "comic" },
        { id: 2, format: "magazine" },
        { id: 3, format: "trade paperback" },
      ];

      render(
        <MarvelContextProvider>
          <Router>
            <Categories formats={formatList} />
          </Router>
        </MarvelContextProvider>
      );

      const comicButton = screen.getByRole("button", { name: "COMIC" });
      const magazineButton = screen.getByRole("button", { name: "MAGAZINE" });
      const paperBackButton = screen.getByRole("button", {
        name: "TRADE PAPERBACK",
      });

      expect(comicButton).toBeInTheDocument();
      expect(magazineButton).toBeInTheDocument();
      expect(paperBackButton).toBeInTheDocument();
    });
    test("Then it should render buttons with formatList info inside", () => {
      const formatList = [
        { id: 1, format: "comic" },
        { id: 2, format: "magazine" },
        { id: 3, format: "trade paperback" },
      ];

      const categoriesComponent = ReactTestRenderer.create(
        <MarvelContextProvider>
          <Categories formats={formatList} />
        </MarvelContextProvider>
      );

      expect(categoriesComponent.toJSON()).toMatchSnapshot();
    });
  });
  /*   describe("When it receives an array of categories and", () => {
    test("Then it should call", () => {
      const { result } = renderHook(() => useMarvel());
      const marvelContext = {
        offset: 20,
        setOffset: jest.fn(),
        limit: 20,
        orderBy: "title",
        setTitle: jest.fn(),
        setComics: [],
        setTotal: 0,
        setComicData: [],
      };

      const formatList = [
        { id: 0, format: "all comics" },
        { id: 1, format: "comic" },
        { id: 2, format: "magazine" },
        { id: 3, format: "trade paperback" },
      ];

      render(
        <MarvelContext.Provider value={marvelContext}>
          <Router>
            <Categories formats={formatList} />
          </Router>
        </MarvelContext.Provider>
      );

      const allComicsButton = screen.getByRole("button", {
        name: "ALL COMICS",
      });

      fireEvent.click(allComicsButton);

      expect(result.current.getComics).toHaveBeenCalled();
    });
  }); */
});
