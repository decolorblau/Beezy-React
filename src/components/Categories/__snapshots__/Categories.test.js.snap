// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the component Categories When it receives an array of categories Then it should render buttons with formatList info inside 1`] = `
<div
  className="formats"
>
  <button
    className="format"
    onClick={[Function]}
  >
    COMIC
  </button>
  <button
    className="format"
    onClick={[Function]}
  >
    MAGAZINE
  </button>
  <button
    className="format"
    onClick={[Function]}
  >
    TRADE PAPERBACK
  </button>
</div>
`;
